File: app/src/am_sonhk/com/example/asian/MyDiffUtilCallback.java
package com.example.asian;

import androidx.recyclerview.widget.DiffUtil;

import com.example.asian.User;

import java.util.List;

public class MyDiffUtilCallback extends DiffUtil.Callback {

    private final List<User> mOldUsers;
    private final List<User> mNewUsers;

    public MyDiffUtilCallback(List<User> oldUsers, List<User> newUsers) {
        this.mOldUsers = oldUsers;
        this.mNewUsers = newUsers;
    }

    @Override
    public int getOldListSize() {
        return mOldUsers.size();
    }

    @Override
    public int getNewListSize() {
        return mNewUsers.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        return mOldUsers.get(oldItemPosition).getmIdUser().equals(mNewUsers.get(newItemPosition).getmIdUser());
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        User oldUser = mOldUsers.get(oldItemPosition);
        User newUser = mNewUsers.get(newItemPosition);
        return oldUser.equals(newUser);
    }
}

File: app/src/am_sonhk/com/example/asian/MyDiffUtilCallback.kt
package com.example.asian

import androidx.recyclerview.widget.DiffUtil
import com.example.asian.model.User

class MyDiffUtilCallback(oldUsers: List<User>, newUsers: List<User>) : DiffUtil.Callback() {
    private val mOldUsers: List<User>
    private val mNewUsers: List<User>

    init {
        mOldUsers = oldUsers
        mNewUsers = newUsers
    }

    override fun getOldListSize(): Int {
        return mOldUsers.size
    }

    override fun getNewListSize(): Int {
        return mNewUsers.size
    }

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return mOldUsers[oldItemPosition].userId
            .equals(mNewUsers[newItemPosition].userId)
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        val oldUser: User = mOldUsers[oldItemPosition]
        val newUser: User = mNewUsers[newItemPosition]
        return oldUser.equals(newUser)
    }
}